File: packages/server/src/routes/git.ts
import type { FastifyInstance } from "fastify";
import type { ServerContext } from "../context.js";
import {
  gitPush,
  listWorktrees,
  addWorktree,
  removeWorktree,
  listBranches,
  updateBranch,
  updateAllBranches,
  type WorktreeAddOptions,
  type ProjectConfig,
} from "@dev-hub/core";
import { resolve } from "node:path";

function validateProjectPath(project: ProjectConfig, workspaceRoot: string): void {
  const root = workspaceRoot.endsWith("/") ? workspaceRoot : workspaceRoot + "/";
  const resolved = resolve(workspaceRoot, project.path);
  if (resolved !== workspaceRoot && !resolved.startsWith(root)) {
    throw new Error(`Project "${project.name}" path escapes workspace root: ${project.path}`);
  }
}

export function registerGitRoutes(app: FastifyInstance, ctx: ServerContext): void {
  function getCtx() {
    if (!ctx.current) throw new Error("No workspace loaded");
    return ctx.current;
  }

  const inProgress = new Set<string>();
  function guard<T>(key: string, fn: () => Promise<T>): Promise<T> {
    if (inProgress.has(key)) {
      throw Object.assign(new Error(`Operation already in progress: "${key}"`), { code: "CONFLICT" });
    }
    inProgress.add(key);
    return fn().finally(() => inProgress.delete(key));
  }

  // POST /api/git/fetch
  app.post<{ Body: { projects?: string[] } }>("/api/git/fetch", async (request, reply) => {
    try {
      const c = getCtx();
      const projects = request.body?.projects?.length
        ? c.config.projects.filter((p) => request.body.projects!.includes(p.name))
        : c.config.projects;
      const result = await guard("fetch", () => c.bulkGitService.fetchAll(projects));
      return reply.send(result);
    } catch (e) {
      return reply.status(500).send({ error: String(e) });
    }
  });

  // POST /api/git/pull
  app.post<{ Body: { projects?: string[] } }>("/api/git/pull", async (request, reply) => {
    try {
      const c = getCtx();
      const projects = request.body?.projects?.length
        ? c.config.projects.filter((p) => request.body.projects!.includes(p.name))
        : c.config.projects;
      const result = await guard("pull", () => c.bulkGitService.pullAll(projects));
      return reply.send(result);
    } catch (e) {
      return reply.status(500).send({ error: String(e) });
    }
  });

  // POST /api/git/push
  app.post<{ Body: { project?: string } }>("/api/git/push", async (request, reply) => {
    try {
      const c = getCtx();
      const projectName = request.body?.project;
      if (!projectName) return reply.status(400).send({ error: "project is required" });
      const project = c.config.projects.find((p) => p.name === projectName);
      if (!project) return reply.status(404).send({ error: `Project "${projectName}" not found` });
      validateProjectPath(project, c.workspaceRoot);
      const result = await guard(`push:${projectName}`, () =>
        gitPush(project.path, project.name, c.bulkGitService.emitter),
      );
      return reply.send(result);
    } catch (e) {
      return reply.status(500).send({ error: String(e) });
    }
  });

  // GET /api/git/:project/worktrees
  app.get<{ Params: { project: string } }>("/api/git/:project/worktrees", async (request, reply) => {
    try {
      const c = getCtx();
      const project = c.config.projects.find((p) => p.name === request.params.project);
      if (!project) return reply.status(404).send({ error: "Project not found" });
      validateProjectPath(project, c.workspaceRoot);
      return reply.send(await listWorktrees(project.path));
    } catch (e) {
      return reply.status(500).send({ error: String(e) });
    }
  });

  // POST /api/git/:project/worktrees
  app.post<{ Params: { project: string }; Body: WorktreeAddOptions }>(
    "/api/git/:project/worktrees",
    async (request, reply) => {
      try {
        const c = getCtx();
        const project = c.config.projects.find((p) => p.name === request.params.project);
        if (!project) return reply.status(404).send({ error: "Project not found" });
        validateProjectPath(project, c.workspaceRoot);
        return reply.send(await addWorktree(project.path, request.body));
      } catch (e) {
        return reply.status(500).send({ error: String(e) });
      }
    },
  );

  // DELETE /api/git/:project/worktrees
  app.delete<{ Params: { project: string }; Body: { path: string } }>(
    "/api/git/:project/worktrees",
    async (request, reply) => {
      try {
        const c = getCtx();
        const project = c.config.projects.find((p) => p.name === request.params.project);
        if (!project) return reply.status(404).send({ error: "Project not found" });
        validateProjectPath(project, c.workspaceRoot);
        await removeWorktree(project.path, request.body.path);
        return reply.send({ removed: true });
      } catch (e) {
        return reply.status(500).send({ error: String(e) });
      }
    },
  );

  // GET /api/git/:project/branches
  app.get<{ Params: { project: string } }>("/api/git/:project/branches", async (request, reply) => {
    try {
      const c = getCtx();
      const project = c.config.projects.find((p) => p.name === request.params.project);
      if (!project) return reply.status(404).send({ error: "Project not found" });
      validateProjectPath(project, c.workspaceRoot);
      return reply.send(await listBranches(project.path));
    } catch (e) {
      return reply.status(500).send({ error: String(e) });
    }
  });

  // POST /api/git/:project/branches/update
  app.post<{ Params: { project: string }; Body: { branch?: string } }>(
    "/api/git/:project/branches/update",
    async (request, reply) => {
      try {
        const c = getCtx();
        const projectName = request.params.project;
        const project = c.config.projects.find((p) => p.name === projectName);
        if (!project) return reply.status(404).send({ error: "Project not found" });
        validateProjectPath(project, c.workspaceRoot);
        const branch = request.body?.branch;
        const result = await guard(`updateBranch:${projectName}`, async () => {
          if (branch) {
            const r = await updateBranch(project.path, branch);
            return [r];
          }
          return updateAllBranches(project.path, c.bulkGitService.emitter);
        });
        return reply.send(result);
      } catch (e) {
        return reply.status(500).send({ error: String(e) });
      }
    },
  );
}

File: server/src/api/git.rs
use axum::{
    Json,
    extract::{Path, State},
    response::IntoResponse,
};
use serde::Deserialize;
use std::path::PathBuf;

use crate::git::{
    BulkGitService, WorktreeAddOptions,
    add_worktree, list_branches, list_worktrees, remove_worktree, update_branch, get_log,
};
use crate::git::bulk::ProjectRef;
use crate::git::progress::create_progress_channel;
use crate::pty::EventSink as _;
use crate::state::AppState;

use super::error::ApiError;

// ---------------------------------------------------------------------------
// POST /api/git/fetch  { projects?: string[] }
// ---------------------------------------------------------------------------

#[derive(Deserialize)]
pub struct ProjectsBody {
    pub projects: Option<Vec<String>>,
}

pub async fn fetch_projects(
    State(state): State<AppState>,
    Json(body): Json<ProjectsBody>,
) -> Result<impl IntoResponse, ApiError> {
    let project_list = collect_project_list(&state, body.projects.as_deref()).await;
    let ssh_cred = state.ssh_creds.read().await.clone();

    // BulkGitService is intentionally per-request: each request manages its own
    // concurrency (Semaphore(4)). Two concurrent fetch requests each run 4 ops = 8 total,
    // which is acceptable for a local dev tool targeting a few dozen projects.
    let bulk = BulkGitService::new(4).with_creds(ssh_cred);
    forward_progress_events(bulk.subscribe(), state.event_sink.clone());

    let refs: Vec<ProjectRef<'_>> = project_list.iter()
        .map(|(n, p)| ProjectRef { name: n.as_str(), path: p.as_path() })
        .collect();
    let results = bulk.fetch_all(&refs).await;
    Ok(Json(results))
}

// ---------------------------------------------------------------------------
// POST /api/git/pull  { projects?: string[] }
// ---------------------------------------------------------------------------

pub async fn pull_projects(
    State(state): State<AppState>,
    Json(body): Json<ProjectsBody>,
) -> Result<impl IntoResponse, ApiError> {
    let project_list = collect_project_list(&state, body.projects.as_deref()).await;
    let ssh_cred = state.ssh_creds.read().await.clone();

    // Intentionally per-request — see fetch_projects for rationale.
    let bulk = BulkGitService::new(4).with_creds(ssh_cred);
    forward_progress_events(bulk.subscribe(), state.event_sink.clone());

    let refs: Vec<ProjectRef<'_>> = project_list.iter()
        .map(|(n, p)| ProjectRef { name: n.as_str(), path: p.as_path() })
        .collect();
    let results = bulk.pull_all(&refs).await;
    Ok(Json(results))
}

// ---------------------------------------------------------------------------
// POST /api/git/push  { project: string }
// ---------------------------------------------------------------------------

#[derive(Deserialize)]
pub struct PushBody {
    pub project: String,
}

pub async fn push_project(
    State(state): State<AppState>,
    Json(body): Json<PushBody>,
) -> Result<impl IntoResponse, ApiError> {
    let path = resolve_project_path(&state, &body.project).await?;
    let progress = Some(create_progress_channel());

    if let Some(ref tx) = progress {
        let mut rx = tx.subscribe();
        let sink = state.event_sink.clone();
        tokio::spawn(async move {
            while let Ok(evt) = rx.recv().await {
                let payload = serde_json::to_value(&evt).unwrap_or_default();
                sink.broadcast("git:progress", payload);
            }
        });
    }

    let result = crate::git::cli_fallback::push(&path, &body.project, &progress).await;
    Ok(Json(result))
}

// ---------------------------------------------------------------------------
// GET /api/git/:project/worktrees
// ---------------------------------------------------------------------------

pub async fn get_worktrees(
    State(state): State<AppState>,
    Path(project): Path<String>,
) -> Result<impl IntoResponse, ApiError> {
    let path = resolve_project_path(&state, &project).await?;
    let worktrees = list_worktrees(&path).await.map_err(ApiError::from_app)?;
    Ok(Json(worktrees))
}

// ---------------------------------------------------------------------------
// POST /api/git/:project/worktrees
// ---------------------------------------------------------------------------

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AddWorktreeBody {
    pub branch: String,
    pub path: Option<String>,
    pub create_branch: Option<bool>,
    pub base_branch: Option<String>,
}

pub async fn add_worktree_route(
    State(state): State<AppState>,
    Path(project): Path<String>,
    Json(body): Json<AddWorktreeBody>,
) -> Result<impl IntoResponse, ApiError> {
    let path = resolve_project_path(&state, &project).await?;
    let opts = WorktreeAddOptions {
        branch: body.branch,
        path: body.path,
        create_branch: body.create_branch.unwrap_or(false),
        base_branch: body.base_branch,
    };
    let worktree = add_worktree(&path, opts).await.map_err(ApiError::from_app)?;
    Ok(Json(worktree))
}

// ---------------------------------------------------------------------------
// DELETE /api/git/:project/worktrees  { path: string }
// ---------------------------------------------------------------------------

#[derive(Deserialize)]
pub struct RemoveWorktreeBody {
    pub path: String,
}

pub async fn remove_worktree_route(
    State(state): State<AppState>,
    Path(project): Path<String>,
    Json(body): Json<RemoveWorktreeBody>,
) -> Result<impl IntoResponse, ApiError> {
    let project_path = resolve_project_path(&state, &project).await?;
    remove_worktree(&project_path, &body.path).await.map_err(ApiError::from_app)?;
    Ok(Json(serde_json::json!({ "ok": true })))
}

// ---------------------------------------------------------------------------
// GET /api/git/:project/branches
// ---------------------------------------------------------------------------

pub async fn get_branches(
    State(state): State<AppState>,
    Path(project): Path<String>,
) -> Result<impl IntoResponse, ApiError> {
    let path = resolve_project_path(&state, &project).await?;
    let branches = list_branches(&path).map_err(ApiError::from_app)?;
    Ok(Json(branches))
}

// ---------------------------------------------------------------------------
// GET /api/git/:project/log
// ---------------------------------------------------------------------------

#[derive(Deserialize)]
pub struct GetLogQuery {
    pub limit: Option<usize>,
}

pub async fn get_log_route(
    State(state): State<AppState>,
    Path(project): Path<String>,
    axum::extract::Query(query): axum::extract::Query<GetLogQuery>,
) -> Result<impl IntoResponse, ApiError> {
    let path = resolve_project_path(&state, &project).await?;
    let limit = query.limit.unwrap_or(100);
    let log = get_log(&path, limit).map_err(ApiError::from_app)?;
    Ok(Json(log))
}

// ---------------------------------------------------------------------------
// POST /api/git/:project/branches/update  { branch?: string }
// ---------------------------------------------------------------------------

#[derive(Deserialize)]
pub struct UpdateBranchBody {
    pub branch: Option<String>,
}

pub async fn update_branch_route(
    State(state): State<AppState>,
    Path(project): Path<String>,
    Json(body): Json<UpdateBranchBody>,
) -> Result<impl IntoResponse, ApiError> {
    let path = resolve_project_path(&state, &project).await?;
    let branch = body.branch.as_deref().unwrap_or("main");
    let result = update_branch(&path, branch, "origin").map_err(ApiError::from_app)?;
    Ok(Json(result))
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

async fn resolve_project_path(state: &AppState, project_name: &str) -> Result<PathBuf, ApiError> {
    state.project_path(project_name).await.map_err(ApiError::from_app)
}

async fn collect_project_list(state: &AppState, filter: Option<&[String]>) -> Vec<(String, PathBuf)> {
    let cfg = state.config.read().await;
    cfg.projects
        .iter()
        .filter(|p| filter.map(|f| f.iter().any(|n| n == &p.name)).unwrap_or(true))
        .map(|p| (p.name.clone(), PathBuf::from(&p.path)))
        .collect()
}

fn forward_progress_events(
    rx: Option<crate::git::progress::ProgressReceiver>,
    sink: crate::pty::BroadcastEventSink,
) {
    if let Some(mut rx) = rx {
        tokio::spawn(async move {
            while let Ok(evt) = rx.recv().await {
                let payload = serde_json::to_value(&evt).unwrap_or_default();
                sink.broadcast("git:progress", payload);
            }
        });
    }
}
